text/0249-v2-experiments.md

```diff
@@ -358,17 +370,288 @@ Disadvantages:
+by `1.61.0-experimental` since the minor part is not equal. This becomes even
+more problematic in npm-v7, which automatically installs `peerDependencies`.
+With npm-v7, executing `npm install` will throw an exception when a consumer
+application declares a dependency on `aws-cdk-lib` version
+`1.61.0-experimental`, tries to consumes a library which declares a
+`peerDependency` on version `^1.60.0`. exception:
+
+`npm install` error:
+
+```
+npm ERR! code ERESOLVE
+npm ERR! ERESOLVE unable to resolve dependency tree
+npm ERR!
+npm ERR! While resolving: my-cdk-application@1.0.0
+npm ERR! Found: aws-cdk-lib@2.3.1-experimental
+npm ERR! node_modules/aws-cdk-lib
+npm ERR!   aws-cdk-lib@"2.3.1-experimental" from the root project
+npm ERR!
+npm ERR! Could not resolve dependency:
+npm ERR! peer aws-cdk-li@"^2.3.0" from my-construct-lib@1.6.0
+npm ERR! node_modules/my-construct-lib
+npm ERR!   my-construct-lib@"1.6.0" from the root project
+npm ERR!
+npm ERR! Fix the upstream dependency conflict, or retry
+npm ERR! this command with --force, or --legacy-peer-deps
+npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
+npm ERR!
+```
+
+Users can work around this by executing `npm install --force` but this is not a
+behavior we want to recommend.
+
+### Option 6: name experimental APIs with Experimental suffix/prefix
+
+The two major changes in v2, monolithic packaging, and no more breaking changes,
+share the same goal - encourage customers to declare a flexible version
+(`^`,`~`) on `aws-cdk-lib`. Fixed versions are a bad practice, they prevent
+users from getting updates that includes critical fixes, security patches and
+more. The mechanism we choose to implement experimental features should adhere
+to the same standard. This means that solutions such as releasing experimental
+features under a different module - `aws-cdk-lib-experimental`, or a different
+version line - `2.x.x-experimental` are not viable solutions. They might prevent
+users from declaring a fixed dependency on the `aws-cdk-lib` mainline release,
+but they mandate it in other release line we own and maintain.
+
+To allow the CDK team to keep experimenting while still adhering to
+[Semantic Versioning](https://semver.org/#semantic-versioning-200), we suggest
+experimental features will be a part of `aws-cdk-lib`, and maintain backward
+compatibility. Experimental features will be add to `aws-cdk-lib` using a
+specific naming scheme. For example, the first experimental version of the
+`grantWrite` API will be introduced under the name `grantWriteExp_v1`, when it's
+time to introduce a breaking change to the API, `grantWriteExp_v1` will be
+deprecated and replaced with a newer version - `grantWriteExp_v2`. When the API
+is finalize it will be introduce as `grantWrite`. Deprecated experimental
+versions will only be removed between major versions release.
+
+#### Implementation
+
+Experimental APIs specification in the AWS CDK docs:
```

Comment:
From a positioning (and also tooling) perspective, I'd frame this like this:

"Experimental APIs will not longer be allowed in aws-cdk-lib. All modules and experimental members will become stable starting 2.0.0. This means that from this release, we are committed to never introduce breaking changes in a non-major bump".

"In parallel, we introduce a new model called "api previews" [or some other name we come with]. This model will explicitly indicate if APIs are not final by including a postfix......."




I'd recommend to write this section as a "working backwards" so it's very easy to understand the new user experience.

Reply:
This is the section we will add to our general documentation, not the v2 release notes. If it will make it clearer I can replace it

Reply:
> "In parallel, we introduce a new model called "api previews" [or some other name we come with]. This model will explicitly indicate if APIs are not final by including a postfix......."
> 


I'm not sure I understand what you mean here. What module are you referring to in "introduce a module"? The way I think about it, an experimental APIs can be in every module. As in the example I listed, S3 can have an experimental API on the bucket construct.

Reply:
_model_ not _module_..


What I am trying to say is that the model and mechanism we have to say for experimentation which is supported by jsii should stay untouched, only that we will no longer allow it in the CDK. So from the users perspective _there is no more "experimentation"_.

There will be a "new thing" with a new name and perhaps a set of new mechanisms to support it and that's the "api previews" or "api stage" or whatever we decide to call it.

Reply:
I think I get what you are saying, you want to completely remove the use of the term "experimental" in v2. v2 is completely stable, and we are adding a new concept "api lifecycle" (trying something), which will allow us to get new APIs early in front of developers so we can receive feedback before we finalize them.

@eladb  is that correct?

Reply:
Replaced the implementation section with the section from v2 release notes.